[ASIM-5600] Read History Matching results

ro-oliveira95 opened the pull request:
Add History Matching metadata and reading functions to aggregator to allow reading HM results data.

ASIM-5600

nicoddemus commented on src/alfasim_sdk/result_reader/aggregator.py:
```diff
@@ -1615,9 +1703,87 @@ def read_global_sensitivity_coefficients(
+    :return:
+        The result HDF file, or None if it doesn't exist or is still being created.
+    """
+    filepath = result_directory / result_filename
+    ignored_file = result_directory / f"{result_filename}.creating"
+
+    if not filepath.is_file():
+        yield None
+    # Avoid to read result files with incomplete metadata.
+    elif not ignored_file.is_file():
+        with _open_result_file(filepath) as file:
+            yield file
+    else:
+        yield None
```
Minor, but I would keep each condition in their own `if`:

```suggestion
    filepath = result_directory / result_filename
    creating_file = result_directory / f"{result_filename}.creating"

    if not filepath.is_file():
        yield None
    # Do not open incomplete files.
    if creating_file.is_file():
        yield None
    else:
        with _open_result_file(filepath) as file:
            yield file
```

The way it is written is a bit more convoluted than it needs to be.

nicoddemus commented on tests/conftest.py:
```diff
@@ -259,3 +269,100 @@ def global_sa_results_dir(datadir: Path) -> Path:
+    dataset[:] = result
+
+    file.swmr_mode = True
+    file.close()
```
Always mind explicit "close" calls like this: this will never be reached if an exception happens until we get to this point.

You should use a context manager or try/finally to ensure the file gets closed even if an exception happens.

ro-oliveira95 replied:
Good catch.

nicoddemus commented on tests/conftest.py:
```diff
@@ -259,3 +269,100 @@ def global_sa_results_dir(datadir: Path) -> Path:
+@pytest.fixture()
+def hm_deterministic_results_dir(datadir: Path) -> Path:
+    """
+    Crete a History Matching result folder with a populated result file for each type of analysis
```
```suggestion
    Create a History Matching result folder with a populated result file for each type of analysis
```

But the doc in general is incorrect, it is only creating one type of analysis here.

ro-oliveira95 replied:
You're correct. I missed this one while refactoring things.

nicoddemus commented on tests/conftest.py:
```diff
@@ -259,3 +269,100 @@ def global_sa_results_dir(datadir: Path) -> Path:
+@pytest.fixture()
+def hm_probabilistic_results_dir(datadir: Path) -> Path:
+    """
+    Crete a History Matching result folder with a populated result file for each type of analysis
```
```suggestion
    Create a History Matching result folder with a populated result file for each type of analysis
```

But the doc in general is incorrect, it is only creating one type of analysis here.

ro-oliveira95 commented:
Ops, I accidentally merged this PR without addressing the review appointments.
Will address then in the base branch's PR https://github.com/ESSS/alfasim-sdk/pull/369.

nicoddemus commented:
> Will address then in the base branch's PR https://github.com/ESSS/alfasim-sdk/pull/369.

Not sure that's what you meant, but please create a new PR targeting #369.
